nixos/lib/testing/run.nix

```
@@ -33,7 +33,8 @@ in
       derivation = hostPkgs.stdenv.mkDerivation {
         name = "vm-test-run-${config.name}";
 
-        requiredSystemFeatures = [ "kvm" "nixos-test" ];
+        requiredSystemFeatures = [ "nixos-test" ]
+          ++ lib.optionals hostPkgs.stdenv.hostPlatform.isLinux [ "kvm" ];
```

I would not assume that virtualization is available on macOS.
Yes, many installations are bare metal, but it's perfectly legal to run it as a guest on Apple hardware, and running a build farm this way offers much more control.
If nested virtualization always works, I'd be happy to leave this flag out, but otherwise we should have a flag like `"hvf"`.

See also
- https://github.com/NixOS/nixpkgs/pull/193336

---

Nix PR:
- https://github.com/NixOS/nix/pull/9187

---

Yes this seems much better. However, it will take some time until that change is normally available on macos machines.... can we make this a follow up in some months?

---

Nested virtualization doesn't work if your run MacOS using `Virtualization.framework` instead of `hvf`. I also am pretty sure nested virtualization is not supported at all yet on M1 and M2

---

> time until that change is normally available on macos machines

It should be in the upcoming stable release, so I'd rather get it right from the start.

If someone doesn't want to update for some reason, they can add the system feature manually.

```suggestion
          ++ lib.optionals hostPkgs.stdenv.hostPlatform.isLinux [ "kvm" ]
          ++ lib.optionals hostPkgs.stdenv.hostPlatform.isDarwin [ "apple-virt" ];
```